modules/tftp/server.rb

```diff
@@ -90,21 +107,93 @@ def pxe_default
+      logger.debug "TFTP:   * Operating system: #{os}"
+      logger.debug "TFTP:   * Version: #{version}"
+      logger.debug "TFTP:   * Architecture: #{arch}"
+      logger.debug "TFTP: Checking bootloader universe if \"#{bootloader_path}\" exists."
+      unless Dir.exist?(bootloader_path)
+        logger.debug "TFTP: Directory \"#{bootloader_path}\" does not exist."
+
+        bootloader_path = "#{bootloader_universe}/pxegrub2/#{os}/default/#{arch}"
+        logger.debug "TFTP: Checking if fallback directory at \"#{bootloader_path}\" exists."
+        unless Dir.exist?(bootloader_path)
+          logger.debug "TFTP: Directory \"#{bootloader_path}\" does not exist."
+          return
+        end
+      end
+
+      bootloader_path
+    end
+
+    def setup_bootloader(mac, os, major, minor, arch, bootfilename_efi, build)
+      pxeconfig_dir_mac = pxeconfig_dir(mac)
+
+      if build
+        logger.debug "TFTP: Host is in build mode."
+        logger.debug "TFTP:   => Deploying host specific bootloader files to \"#{pxeconfig_dir_mac}\""
+
+        FileUtils.mkdir_p(pxeconfig_dir_mac)
+
+        version = (minor && !minor.empty?) ? "#{major}.#{minor}" : major
+        bootloader_path = bootloader_path(os, version, arch)
+
+        if bootloader_path
+          logger.debug "TFTP: Copying bootloader files from bootloader universe:"
+          logger.debug "TFTP:   - \"#{bootloader_path}/*\" => \"#{pxeconfig_dir_mac}/\""
+          FileUtils.cp_r("#{bootloader_path}/.", "#{pxeconfig_dir_mac}/", remove_destination: true)
+        else
+          pxeconfig_dir = pxeconfig_dir()
+
+          logger.debug "TFTP: Copying default bootloader files:"
+          logger.debug "TFTP:   - \"#{pxeconfig_dir}/grub#{bootfilename_efi}.efi\" => \"#{pxeconfig_dir_mac}/grub#{bootfilename_efi}.efi\""
+          logger.debug "TFTP:   - \"#{pxeconfig_dir}/shim#{bootfilename_efi}.efi\" => \"#{pxeconfig_dir_mac}/shim#{bootfilename_efi}.efi\""
+          logger.debug "TFTP:   - \"#{pxeconfig_dir}/grub#{bootfilename_efi}.efi\" => \"#{pxeconfig_dir_mac}/boot.efi\""
+          logger.debug "TFTP:   - \"#{pxeconfig_dir}/shim#{bootfilename_efi}.efi\" => \"#{pxeconfig_dir_mac}/boot-sb.efi\""
+          FileUtils.cp_r("#{pxeconfig_dir}/grub#{bootfilename_efi}.efi", "#{pxeconfig_dir_mac}/grub#{bootfilename_efi}.efi", remove_destination: true)
+          FileUtils.cp_r("#{pxeconfig_dir}/shim#{bootfilename_efi}.efi", "#{pxeconfig_dir_mac}/shim#{bootfilename_efi}.efi", remove_destination: true)
+          FileUtils.cp_r("#{pxeconfig_dir}/grub#{bootfilename_efi}.efi", "#{pxeconfig_dir_mac}/boot.efi", remove_destination: true)
+          FileUtils.cp_r("#{pxeconfig_dir}/shim#{bootfilename_efi}.efi", "#{pxeconfig_dir_mac}/boot-sb.efi", remove_destination: true)
+        end
+
+        File.write(File.join(pxeconfig_dir_mac, 'os_info'), "#{os} #{version} #{arch}")
+      else
+        logger.debug "TFTP: Host is not in build mode."
```

Comment:
Everytime I create new host, I'll see this message in the log.

```
2024-06-27T14:47:21 d924fd47 [D] TFTP: Host is not in build mode.
2024-06-27T14:47:21 d924fd47 [D] TFTP:   => Removing host specific bootloader files from "/var/lib/tftpboot//host_config/00-aa-bb-6b-06-dd/grub2"
2024-06-27T14:47:21 d924fd47 [D] TFTP: /var/lib/tftpboot//host_config/00-aa-bb-6b-06-dd/grub2/grub.cfg created successfully
2024-06-27T14:47:21 d924fd47 [D] TFTP: /var/lib/tftpboot//host_config/00-aa-bb-6b-06-dd/grub2/grub.cfg-01-00-aa-bb-6b-06-dd created successfully
2024-06-27T14:47:21 d924fd47 [D] TFTP: /var/lib/tftpboot//host_config/00-aa-bb-6b-06-dd/grub2/grub.cfg-00:aa:bb:6b:06:dd created successfully
2024-06-27T14:47:21 d924fd47 [D] TFTP: /var/lib/tftpboot//grub2/grub.cfg-01-00-aa-bb-6b-06-dd created successfully
2024-06-27T14:47:21 d924fd47 [D] TFTP: /var/lib/tftpboot//grub2/grub.cfg-00:aa:bb:6b:06:dd created successfully
2024-06-27T14:47:21 d924fd47 [I] Finished POST /tftp/PXEGrub2/00:aa:bb:6b:06:dd with 200 (1.59 ms)
```

Why is it needed?

Reply:
I don't see it. When I create a new host, I see the `Host is in build mode.` message in the log. However, I would expect to see this message (`Host is not in build mode.`) when the host is not in build mode (if I cancel it manually, or after the provisioning ends for this host). Instead, I get the `Host is in build mode.` message regardless. This wasn't the case in the past; I think the recent changes are causing this issue.

Reply:
Hey, I tried to verify it and I can confirm, this worked differently before (at least to the best of my knowledge and my notes :sweat_smile:). :thinking:

> I think the recent changes are causing this issue.

I have to further investigate that and have to take a look at the history of the PR.
The initial idea of the "delete the host specific bootloader files in case `build` is `false`" was to remove the existing bootloader files in case a host is rebuilt. But my current guess is that there is in general no need to rely on the `build` flag for that. It would be a way more atomic (and simpler) approach to simply delete the existing host specific bootloader files before deploying new bootloader files in the host specific directory. But I have to verify all of this.

Unfortunately today is my day off, therefor I will have to take a closer look on Monday. :see_no_evil:

Reply:
Ok, now I am wiser :sweat_smile:

Indeed it was intentional [to delete the bootloader files when the host leaves Build Mode](https://community.theforeman.org/t/add-secureboot-support-for-arbitrary-distributions/32601/30)  (i.e. when provisioning is finished) but with the knowledge I gained in the course of these PRs this doesn't make sense at all as also the provisioned hosts require the files in the `host_config/<MAC>/grub2` directory to boot.
So it seems that the deletion of the bootloader files has never worked as it was intended to as then no host that uses any `Grub2 UEFI*` PXE loader could have been rebooted after being deployed.

I propose to remove the deletion depending on the build flag, but instead to delete the files every time before they get deployed (i.e. every time the TFTP configuration is deployed). By this we guarantee a consistent state of the files in the Bootloader Universe and in the `host_config/<MAC>/grub2` directories.

Reply:
Done :)

Reply:
> ...this doesn't make sense at all as also the provisioned hosts require the files in the `host_config/<MAC>/grub2` directory to boot. So it seems that the deletion of the bootloader files has never worked as it was intended to as then no host that uses any `Grub2 UEFI*` PXE loader could have been rebooted after being deployed.

Why do the provisioned hosts require the files in the `host_config/<MAC>/grub2` directory to boot? 
The provisioned hosts should boot from the local hard drive and not go through the entire provisioning process each time they are booted.

Note: During the boot process, by default the machine first tries to boot from the network (i.e., it tries to download the boot*.efi files). However, even if these files are missing, the boot process will still be successful because it will fall back to booting from the local hard drive.

Reply:
I'll try to summarize what we discussed here:
Since provisioned hosts also need shim/grub to boot, it's necessary to retain these binaries after provisioning. However, this raises concerns about storage usage and manual updates if binaries are updated.

After our DM discussion, we opted for a solution: implementing a dedicated `bootloader_universe` directory within the TFTP root (e.g., `/var/lib/tftpboot/bootloader_universe`). This approach involves using symlinks for shim/grub for each host consistently, as detailed [here](https://community.theforeman.org/t/add-secureboot-support-for-arbitrary-distributions/32601/26?u=jan).

Reasons for this decision include:
1. This maintains the default behavior in Foreman (prior to this PR). Previously, we never deleted the shim/grub.
2. Even with the correct boot order, when a machine boots, it first tries to download the shim/grub. Therefore, we want to keep these files even when the host is not in build mode.
3. Using symlinks instead of copies saves storage space and simplifies maintenance if updates are required.
4. In the event of a certificate revocation, there's a workaround to deploy the correct shim/grub elsewhere and manually relink all affected hosts to ensure functionality.